src/Bicep.Core/Semantics/MetadataSymbol.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Collections.Generic;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Syntax;
+
+namespace Bicep.Core.Semantics
+{
+    public class MetadataSymbol : DeclaredSymbol
+    {
+        public MetadataSymbol(ISymbolContext context, string name, MetadataDeclarationSyntax declaringSyntax, SyntaxBase value)
+            : base(context, name, declaringSyntax, declaringSyntax.Name)
+        {
+            this.Value = value;
+        }
+
+        public SyntaxBase Value { get; }
+
+        public override void Accept(SymbolVisitor visitor) => visitor.VisitMetadataSymbol(this);
+
+        public override SymbolKind Kind => SymbolKind.Metadata;
+
+        public override IEnumerable<Symbol> Descendants
+        {
+            get
+            {
+                yield return this.Type;
+            }
+        }
+
+        public override IEnumerable<ErrorDiagnostic> GetDiagnostics() => RestrictedIdentifierValidatorVisitor.GetDiagnostics(this);
+
+        private sealed class RestrictedIdentifierValidatorVisitor : SymbolVisitor
```

Comment:
Since we are only checking the metadata symbol name, the visitor doesn't seem to be necessary. Could you replace it with a helper method?

Reply:
I hope this is how you intended it, otherwise please let me know.

Reply:
I hope this is how you intended it, otherwise please let me know.